### PR TITLE
fix issue 16245 - the message emitted when a const function mutates members is misleading

### PR DESCRIPTION
**Pull request by ghost**
Indicates clearly if a mutable LValue becomes immutable because it's used in a const or immutable function.

### CODE REVIEW

File: src/expression.d [hunk 1]
```diff
@@ -2778,7 +2778,26 @@ extern (C++) abstract class Expression : RootObject
...
+                    if (isNeedThisScope(sc, (cast(DotVarExp) this).var))
+                        for (Dsymbol s = sc.func; s; s = s.toParent2())
+                    {
+                        FuncDeclaration ff = s.isFuncDeclaration();
+                        if (!ff)
+                            break;
+                        if (!ff.type.isMutable)
+                        {
+                            error("cannot modify %s in %s function", toChars(), MODtoChars(type.mod));
+                            errFromFunc = true;
```
Comment (WalterBright) [inline src/expression.d]: Dispense with `errFromFunc` and just do a `return new ErrorExp()` here.

Comment (BBasile) [inline src/expression.d]: understood, done.

**Comment by BBasile** [comment]
It's just an __autopsy__. The error is here so the compilation will fail. The approach is just try to detect a pattern that lead to the failure, if it's detected then put it on stderr, if not then keep the error message from the compiler programmers point of view.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Pull request merged**